### PR TITLE
Add note_full_data parsing from notification payload, open proper order from notification

### LINKED ISSUE #13573: [Orders] Order push notification always opens most recent order
**Describe the bug**

When I receive multiple push notifications for new orders, tapping on a notification always opens the most recent order (not necessarily the order corresponding to the notification I tapped on).

This works as expected for review push notifications, and for order push notifications on Android, so there might be something we can borrow from those implementations to open the correct order.

**To Reproduce**
1. Confirm push notifications are enabled for the app and background the app.
2. On the web, create at least two new orders.
3. On the device, tap the older push notification.
4. Notice the order details open for the most recent order, not the order corresponding to the notification that was tapped.

**Screenshots**

In this screencast I tapped the second notification but it opened the order corresponding to the first notification, instead:

https://github.com/user-attachments/assets/396e6217-3fb2-4636-8d1e-5de7f8bf22a8

**Expected behavior**

I expect that tapping on a push notification will always open the corresponding order.

### PR DESCRIPTION
**Pull request by bozidarsevo**
Closes: #13573

## Description

We noticed that iOS notification payloads were missing `note_full_data` parameter which in a combination of getting the same `note_id` for all new orders notification resulted in a bug where users would tap the new order notification and it would take them to the last order and not the one the notification was for.

Solution was to start sending the `note_full_data` parameter from backend and for client side to parse and use it.

So with this PR the app parses the `note_full_data` parameter (if there is data for it) and if it has data for the specific order we show the order.

The only thing I have left todo is marking the notification as read on watch and in notification extension. Since bunch of code/actions are not available from watch and notif extension I have some issues with that. Can we do that part in a separate PR? @Ecarrion 

## Testing information

Testing should be done in **3 parts**:

1. General notification opening for the iOS app
- Get few new orders for one of the stores you have so it triggers notifications for it. It is important to have more of them so you can open order that is not the last.
- Tap on the notification for the order that was not the last order
- Check that the app opens the proper order
- Tap on the notification for the last order
- Check that the app opens the proper order and that the notification is marked as read also on the web

https://github.com/user-attachments/assets/4e2edf8e-9b73-4244-bb23-d93361f23c6d

2. General notification opening for the watchOS app
- Do the same steps from part 1 but open notifications from your Apple Watch

https://github.com/user-attachments/assets/80865274-ad87-4cee-a45b-ea1f198e3517

3. Notification extension
- Get a new ordera for one of the stores you have so it triggers notification for it.
- Long press the notification and check the notification preview is shown correctly

IMG_723E898202D3-1

## Screenshots

- Videos and screenshots are above with testing instructions

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Woo Watch App/Orders/OrderDetailLoader.swift [hunk 1]
```diff
@@ -66,7 +66,7 @@ final class OrderDetailLoaderViewModel: ObservableObject {
         self.viewState = .loading
         let dataService = OrderNotificationDataService(credentials: dependencies.credentials)
         do {
-            let (_, remoteOrder) = try await dataService.loadOrderFrom(noteID: pushNotification.noteID)
+            let (_, remoteOrder) = try await dataService.loadOrderFrom(notification: pushNotification)
```
Comment (bozidarsevo) [inline WooCommerce/Woo Watch App/Orders/OrderDetailLoader.swift:69]: Passing the entire `pushNotification` so we can do the loading of the order from the `note_full_data` part of the payload

File: WooCommerce/NotificationExtension/OrderNotificationDataService.swift [hunk 2]
```diff
@@ -36,6 +36,19 @@ final class OrderNotificationDataService {
...
+        let note: Note
+        if let notificationNote = notification.note {
+            note = notificationNote
+        }
+        else {
+            note = try await loadNotification(noteID: notification.noteID)
+        }
+        let order = try await loadOrder(from: note)
+        return (note, order)
+    }
```
Comment (bozidarsevo) [inline WooCommerce/NotificationExtension/OrderNotificationDataService.swift]: If there is a `note` (parsed from `note_full_data`) we use it, instead we load the `note` using the `noteID`

File: WooCommerce/Classes/ServiceLocator/PushNotification.swift [hunk 3]
```diff
@@ -65,3 +101,41 @@ extension PushNotification: Identifiable {
...
+            "user": "212589093",
+            "title": "You have a new order! 🎉",
+        ]
+        // Deliver the notification in five seconds.
+        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: 5, repeats: false)
+        let request = UNNotificationRequest(identifier: "FiveSecondNewOrdeTestingNotification", content: content, trigger: trigger) // Schedule the notification.
+        let center = UNUserNotificationCenter.current()
+        center.add(request)
+    }
+}
```
Comment (bozidarsevo) [inline WooCommerce/Classes/ServiceLocator/PushNotification.swift]: Just a helper method for scheduling local notification for easier testing.
I can remove it if that would be better?

**Comment by bozidarsevo** [comment]
- [x] extract note_full_data from the payload
- [x] if needed mark the notification as read (if the notification is for the last order)
- [x] check fetching notifications on the watch
- [x] check fetching notifications on the app extension
- [x] check fetching notifications as background updates

File: WooCommerce/Classes/Model/Note+Woo.swift [hunk 4]
```diff
@@ -66,3 +66,34 @@ private extension Note {
         static let emptyStar    = "\u{2606}"  // Unicode White Star ☆
     }
 }
+
+extension Note {
+    var orderID: Int? {
+        return meta.identifier(forKey: .order)
+    }
+
+    func markOrderNoteAsReadIfNeeded() {
```
Comment (bozidarsevo) [inline WooCommerce/Classes/Model/Note+Woo.swift:77]: Added this helper method to centralize marking note as read.
Basically we check if the order id of this Note and the synced one is the same, if it is we mark it as read.

File: Networking/Networking/Model/Note.swift [hunk 5]
```diff
@@ -238,3 +237,47 @@ extension Note {
...
+                  type: type,
+                  subtype: subtype,
+                  url: url,
+                  title: title,
+                  subject: rawSubjectData ?? Data(),
+                  header: rawHeaderData ?? Data(),
+                  body: rawBodyData ?? Data(),
+                  meta: rawMetaData ?? Data())
+    }
+}
```
Comment (Ecarrion) [inline Networking/Networking/Model/Note.swift]: To not duplicate and maintain this code twice, do you think we could do something like
```swift
let data = try JSONSerialization.data(withJSONObject: payload)
return try JSONDecoder().decode(Note.self, from: data)
```

That way we will only handle serialization in the custom decoding method

File: WooCommerce/Classes/Model/Note+Woo.swift [hunk 6]
```diff
@@ -66,3 +66,34 @@ private extension Note {
...
+               syncronizedNoteOrderID == currentNoteOrderID {
+                // mark as read
+                let syncAction = NotificationAction.updateReadStatus(noteID: noteID, read: true) { error in
+                    DDLogError("⛔️ Error marking single notification as read: \(error)")
+                }
+                ServiceLocator.stores.dispatch(syncAction)
+            }
+        }
+        ServiceLocator.stores.dispatch(action)
+    }
```
Comment (Ecarrion) [inline WooCommerce/Classes/Model/Note+Woo.swift]: While this works, We tend to avoid accessing Yosemite from networking classes. 

One pattern we use frequently is to create a `MarkOrderAsReadUseCase` type where we can properly pass dependencies, like the `stores`, and test it appropriately.

Additionally, having a completion closure or making it async and return a completion value would be nice for future debugging!

File: WooCommerce/Classes/ServiceLocator/PushNotification.swift [hunk 7]
```diff
@@ -4,6 +4,11 @@ import Foundation
 import struct Yosemite.Note
 #elseif canImport(NetworkingWatchOS)
 import struct NetworkingWatchOS.Note
+#elseif canImport(Networking)
+import struct Networking.Note
+#endif
```
Comment (Ecarrion) [inline WooCommerce/Classes/ServiceLocator/PushNotification.swift]: Can you explain a little bit why is this import needed?

If I remember correctly, `Yosemite.Note` is the same as `Networking.Note`

Is it to be able to link in in the extension? If so maybe we should change the initial `Yosemite.Note` to not confuse any other reader 😅

File: WooCommerce/Classes/ServiceLocator/PushNotification.swift [hunk 8]
```diff
@@ -42,7 +50,30 @@ extension PushNotification {
               }
         let subtitle = alert.string(forKey: APNSKey.alertSubtitle)
         let message = alert.string(forKey: APNSKey.alertMessage)
-        return PushNotification(noteID: noteID, siteID: siteID, kind: noteKind, title: title, subtitle: subtitle, message: message)
+        let note: Note? = noteFromNoteData(userInfo.string(forKey: APNSKey.noteFullData))
+        return PushNotification(noteID: noteID, siteID: siteID, kind: noteKind, title: title, subtitle: subtitle, message: message, note: note)
+    }
+
+    static func noteFromNoteData(_ noteFulldata: String?) -> Note? {
```
Comment (Ecarrion) [inline WooCommerce/Classes/ServiceLocator/PushNotification.swift]: Do you think that `NoteFromCompressedData` could make more sense to the reader?
Also, adding some comments on the method signature could help future readers.

File: WooCommerce/Classes/ServiceLocator/PushNotification.swift [hunk 9]
```diff
@@ -65,3 +101,43 @@ extension PushNotification: Identifiable {
         noteID
     }
 }
+
+#if DEBUG
+extension PushNotification {
+    static func scheduleTestingOrderNotification() {
```
Comment (Ecarrion) [inline WooCommerce/Classes/ServiceLocator/PushNotification.swift]: Would this be removed or are you planning to leave it for future testing/debug?

File: WooCommerce/Classes/ViewRelated/MainTabBarController.swift [hunk 10]
```diff
@@ -382,6 +382,17 @@ extension MainTabBarController {
         ServiceLocator.stores.dispatch(action)
     }
 
+    static func switchStoreIfNeededAndPresentNotificationDetails(notification: WooCommerce.PushNotification) {
+        guard let note = notification.note else {
+            presentNotificationDetails(for: notification.noteID)
+            return
+        }
+        let siteID = Int64(note.meta.identifier(forKey: .site) ?? Int.min)
```
Comment (Ecarrion) [inline WooCommerce/Classes/ViewRelated/MainTabBarController.swift]: Should we exist if we don't find `note.meta.identifier(forKey: .site)` as a value?

**Review by Ecarrion — COMMENTED** [review]
Lastly, do you think we could add some unit tests?

I think we could unit test parsing the full data object and the new `loadOrderFrom(notification: PushNotification)`

File: Networking/Networking/Model/Note.swift (hunk 5 above)
Comment (bozidarsevo) [reply Networking/Networking/Model/Note.swift]: Will try it out now, tnx!

File: WooCommerce/Classes/ServiceLocator/PushNotification.swift (hunk 9 above)
Comment (bozidarsevo) [reply WooCommerce/Classes/ServiceLocator/PushNotification.swift]: It could be used for future testing so I have a dilemma if I should remove it or not.

File: WooCommerce/Classes/ViewRelated/MainTabBarController.swift (hunk 10 above)
Comment (bozidarsevo) [reply WooCommerce/Classes/ViewRelated/MainTabBarController.swift]: I basically copied and modified a bit logic from `presentNotificationDetails(for noteID: Int64)` which had this line `let siteID = Int64(note.meta.identifier(forKey: .site) ?? Int.min)`. What do you think, should we return if there is no `siteID`?

File: Networking/Networking/Model/Note.swift (hunk 5 above)
Comment (bozidarsevo) [reply Networking/Networking/Model/Note.swift]: I remembered now that I tried that and the problem are lines like:
`let hash = try container.decode(Int64.self, forKey: .hash)`
Only if I change `public init(from decoder: Decoder) throws {` to decode `hash` if there is some, and use `Int64.min` if there is no `hash` to be decoded? Since notification payload does not have `hash` for the note.

Comment (bozidarsevo) [reply Networking/Networking/Model/Note.swift]:
```
let hashDecoded = try? container.decode(Int64.self, forKey: .hash)
let hash = hashDecoded ?? Int64.min
```

File: WooCommerce/Classes/Model/Note+Woo.swift (hunk 6 above)
Comment (bozidarsevo) [reply WooCommerce/Classes/Model/Note+Woo.swift]: Even though this is in extension for `Note` we try to avoid that?

 I suppose since I only need note and order identifier I will make a method that uses just that. 👍

Comment (bozidarsevo) [reply WooCommerce/Classes/Model/Note+Woo.swift]: > MarkOrderAsReadUseCase

Could you explain that a bit more? That would be some type with just a method that marks the note as read or?

Comment (bozidarsevo) [reply WooCommerce/Classes/Model/Note+Woo.swift]: something like this?
```
class MarkOrderAsReadUseCase {
    static func markOrderNoteAsReadIfNeeded(stores: StoresManager, noteID: Int64, orderID: Int) {
        let action = NotificationAction.synchronizeNotification(noteID: noteID) { syncronizedNote, error in
            guard let syncronizedNote = syncronizedNote else {
                return
            }
            if let syncronizedNoteOrderID = syncronizedNote.orderID,
               syncronizedNoteOrderID == orderID {
                // mark as read
                let syncAction = NotificationAction.updateReadStatus(noteID: noteID, read: true) { error in
                    DDLogError("⛔️ Error marking single notification as read: \(error)")
                }
                stores.dispatch(syncAction)
            }
        }
        stores.dispatch(action)
    }
}

```
I would add a completion block also.

File: WooCommerce/Classes/ServiceLocator/PushNotification.swift (hunk 7 above)
Comment (bozidarsevo) [reply WooCommerce/Classes/ServiceLocator/PushNotification.swift]: Yosemite.Note:
`public typealias Note = Networking.Note`

I think I added it because Yosemite is not available for notification extension?
If you look at `OrderNotificationDataService.swift` it imports `Networking`

File: Networking/Networking/Model/Note.swift (hunk 5 above)
Comment (Ecarrion) [reply Networking/Networking/Model/Note.swift]: That works, we also have `decodeIfPresent` method!

File: WooCommerce/Classes/Model/Note+Woo.swift (hunk 6 above)
Comment (Ecarrion) [reply WooCommerce/Classes/Model/Note+Woo.swift]: I guess it could be a struct, and you could make the methods async to not deal with completion closures.

And maybe you could use compile flags and separate internal implementations to allow this to be shared with the watch & notifications extensions later.

File: WooCommerce/Classes/ServiceLocator/PushNotification.swift (hunk 7 above)
Comment (Ecarrion) [reply WooCommerce/Classes/ServiceLocator/PushNotification.swift]: I think it is safe to change the initial `import Yosemite.Note` to `import Networking.Note`

File: WooCommerce/Classes/ServiceLocator/PushNotification.swift (hunk 9 above)
Comment (Ecarrion) [reply WooCommerce/Classes/ServiceLocator/PushNotification.swift]: I rather test things from real notifications to make sure that everything is working as expected! I would vote to remove it!

File: WooCommerce/Classes/ViewRelated/MainTabBarController.swift (hunk 10 above)
Comment (Ecarrion) [reply WooCommerce/Classes/ViewRelated/MainTabBarController.swift]: I think so, basically as the same guard as above!

**Review by Ecarrion — COMMENTED** [review]
Hand tested it and it seems to be working good! Great job 🎉 

One other thing to update is how we get the notification on the [PushNotificationBackgroundSynchronizer](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/Notifications/PushNotificationBackgroundSynchronizer.swift#L44) to get the order from the payload when possible. But feel free to tackle that in another PR

**Comment by bozidarsevo** [comment]
> But feel free to tackle that in another PR

Yeah, better to do it separate to avoid possible bugs here. :)

File: WooCommerce/WooCommerceTests/Notifications/PushNotificationTests.swift [hunk 11]
```diff
@@ -0,0 +1,52 @@
+//import Combine
+import Experiments
+import XCTest
+import Yosemite
+@testable import WooCommerce
+
+final class PushNotificationTests: XCTestCase {
+    private let userInfo: [AnyHashable: Any] = [
```
Comment (bozidarsevo) [inline WooCommerce/WooCommerceTests/Notifications/PushNotificationTests.swift:6]: Used that payload from testing here for test so we use it for this test and if needed in the future we can use it to copy/paste it if we need to trigger some local notification :)

File: WooCommerce/Classes/Model/Note+Woo.swift [hunk 12]
```diff
@@ -66,3 +66,20 @@ private extension Note {
         static let emptyStar    = "\u{2606}"  // Unicode White Star ☆
     }
 }
+
+extension Note {
+    var orderID: Int? {
+        return meta.identifier(forKey: .order)
+    }
+
+    func markOrderNoteAsReadIfNeeded() {
```
Comment (Ecarrion) [inline WooCommerce/Classes/Model/Note+Woo.swift:77]: Since you opted for this approach, it would be helpful to add some comments to this method indicating that a network request will be performed.

File: WooCommerce/NotificationExtension/OrderNotificationDataService.swift [hunk 13]
```diff
@@ -36,6 +36,19 @@ final class OrderNotificationDataService {
...
 
+    @MainActor
+    func loadOrderFrom(notification: PushNotification) async throws -> (Note, Order) {
+        let note: Note
+        if let notificationNote = notification.note {
+            note = notificationNote
+        }
+        else {
+            note = try await loadNotification(noteID: notification.noteID)
+        }
```
Comment (Ecarrion) [inline WooCommerce/NotificationExtension/OrderNotificationDataService.swift]: Tip: For these kind of code I'd like to do something like 

```swift
let note: Note = try await {
    if let notificationNote = notification.note {
        return notificationNote
    } else {
        return try await loadNotification(noteID: notification.noteID)
    }
}()
```

In order to make it clearer for the reader that the if/else is just for the note creation and should not have any other side effect

**Review by Ecarrion — APPROVED** [review]
Looks great @bozidarsevo! 

One other thing you could do is to add some unit tests to `MarkOrderAsReadUseCase`

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/222">20.3</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by bozidarsevo** [comment]
> One other thing you could do is to add some unit tests to `MarkOrderAsReadUseCase`

Created an issue for it https://github.com/woocommerce/woocommerce-ios/issues/13851 since it needs some other changes in the code base so it can be fully testable
